File: days/day_08/solutions/solution.jl
mutable struct UnionFind
    parent::Vector{Int}
    size::Vector{Int}
    component_count::Int
end

function UnionFind(n::Int)
    UnionFind(collect(1:n), ones(Int, n), n)
end

function find_root!(uf::UnionFind, x::Int)::Int
    if uf.parent[x] != x
        uf.parent[x] = find_root!(uf, uf.parent[x])
    end
    return uf.parent[x]
end

function union!(uf::UnionFind, x::Int, y::Int)::Bool
    root_x = find_root!(uf, x)
    root_y = find_root!(uf, y)
    if root_x == root_y
        return false
    end
    if uf.size[root_x] < uf.size[root_y]
        root_x, root_y = root_y, root_x
    end
    uf.parent[root_y] = root_x
    uf.size[root_x] += uf.size[root_y]
    uf.component_count -= 1
    return true
end

struct Coord
    x::Int
    y::Int
    z::Int
end

function parse_coordinates(lines::Vector{String})::Vector{Coord}
    coords = Coord[]
    for line in lines
        line = strip(line)
        isempty(line) && continue
        parts = split(line, ',')
        x = parse(Int, parts[1])
        y = parse(Int, parts[2])
        z = parse(Int, parts[3])
        push!(coords, Coord(x, y, z))
    end
    return coords
end

function squared_distance(p1::Coord, p2::Coord)::Int
    dx = p2.x - p1.x
    dy = p2.y - p1.y
    dz = p2.z - p1.z
    return dx * dx + dy * dy + dz * dz
end

function solve(input_data::String)::Tuple{String, String}
    lines = split(strip(input_data), '\n') .|> String
    coords = parse_coordinates(lines)
    n = length(coords)
    
    if n == 0
        return ("0", "0")
    end
    
    # Generate all pairs with squared distances
    pairs = Tuple{Int, Int, Int}[]
    for i in 1:n
        for j in (i+1):n
            dist_sq = squared_distance(coords[i], coords[j])
            push!(pairs, (i, j, dist_sq))
        end
    end
    
    # Sort by distance
    sort!(pairs, by = x -> x[3])
    
    # Part 1: Connect first 1000 pairs
    uf1 = UnionFind(n)
    connections_made = 0
    for (i, j, _) in pairs
        if connections_made >= 1000
            break
        end
        union!(uf1, i, j)
        connections_made += 1
    end
    
    # Get component sizes
    component_sizes = Dict{Int, Int}()
    for i in 1:n
        root = find_root!(uf1, i)
        component_sizes[root] = uf1.size[root]
    end
    
    sizes = sort(collect(values(component_sizes)), rev=true)
    part1 = length(sizes) >= 3 ? sizes[1] * sizes[2] * sizes[3] : 0
    
    # Part 2: Connect until all in one circuit
    uf2 = UnionFind(n)
    final_pair = nothing
    for (i, j, _) in pairs
        if uf2.component_count == 1
            break
        end
        if union!(uf2, i, j)
            if uf2.component_count == 1
                final_pair = (i, j)
                break
            end
        end
    end
    
    part2 = if final_pair !== nothing
        i, j = final_pair
        coords[i].x * coords[j].x
    else
        0
    end
    
    return (string(part1), string(part2))
end

function main()
    data = read("../data/input.txt", String) |> strip |> String
    part1, part2 = solve(data)
    println("Part 1: $part1")
    println("Part 2: $part2")
end

main()

File: days/day_08/solutions/solution.rs
use std::fs;
use std::collections::HashMap;

type Coord = (i64, i64, i64);
type Pair = (usize, usize, i64);

struct UnionFind {
    parent: Vec<usize>,
    size: Vec<i64>,
    component_count: usize,
}

impl UnionFind {
    fn new(n: usize) -> Self {
        UnionFind {
            parent: (0..n).collect(),
            size: vec![1; n],
            component_count: n,
        }
    }

    fn find(&mut self, x: usize) -> usize {
        if self.parent[x] != x {
            self.parent[x] = self.find(self.parent[x]);
        }
        self.parent[x]
    }

    fn union(&mut self, x: usize, y: usize) -> bool {
        let root_x = self.find(x);
        let root_y = self.find(y);
        if root_x == root_y {
            return false;
        }
        let (root_x, root_y) = if self.size[root_x] < self.size[root_y] {
            (root_y, root_x)
        } else {
            (root_x, root_y)
        };
        self.parent[root_y] = root_x;
        self.size[root_x] += self.size[root_y];
        self.component_count -= 1;
        true
    }
}

fn parse_coordinates(lines: &[&str]) -> Vec<Coord> {
    lines.iter()
        .filter_map(|line| {
            let trimmed = line.trim();
            if trimmed.is_empty() {
                return None;
            }
            let parts: Vec<&str> = trimmed.split(',').collect();
            if parts.len() != 3 {
                return None;
            }
            let x: i64 = parts[0].parse().ok()?;
            let y: i64 = parts[1].parse().ok()?;
            let z: i64 = parts[2].parse().ok()?;
            Some((x, y, z))
        })
        .collect()
}

fn squared_distance(p1: Coord, p2: Coord) -> i64 {
    let dx = p2.0 - p1.0;
    let dy = p2.1 - p1.1;
    let dz = p2.2 - p1.2;
    dx * dx + dy * dy + dz * dz
}

fn solve(input_data: &str) -> (String, String) {
    let lines: Vec<&str> = input_data.trim().lines().collect();
    let coords = parse_coordinates(&lines);

    let n = coords.len();
    if n == 0 {
        return ("0".to_string(), "0".to_string());
    }

    // Generate all pairs with squared distances
    let mut pairs: Vec<Pair> = Vec::new();
    for i in 0..n {
        for j in (i + 1)..n {
            let dist_sq = squared_distance(coords[i], coords[j]);
            pairs.push((i, j, dist_sq));
        }
    }

    // Sort by distance
    pairs.sort_by_key(|&(_, _, dist_sq)| dist_sq);

    // Part 1: Connect first 1000 pairs
    let mut uf1 = UnionFind::new(n);
    let mut connections_made = 0;
    for &(i, j, _) in &pairs {
        if connections_made >= 1000 {
            break;
        }
        uf1.union(i, j);
        connections_made += 1;
    }

    // Get component sizes
    let mut component_sizes: HashMap<usize, i64> = HashMap::new();
    for i in 0..n {
        let root = uf1.find(i);
        component_sizes.insert(root, uf1.size[root]);
    }

    let mut sizes: Vec<i64> = component_sizes.values().copied().collect();
    sizes.sort_by(|a, b| b.cmp(a));
    let part1 = if sizes.len() >= 3 {
        sizes[0] * sizes[1] * sizes[2]
    } else {
        0
    };

    // Part 2: Connect until all in one circuit
    let mut uf2 = UnionFind::new(n);
    let mut final_pair: Option<(usize, usize)> = None;
    for &(i, j, _) in &pairs {
        if uf2.component_count == 1 {
            break;
        }
        if uf2.union(i, j) {
            if uf2.component_count == 1 {
                final_pair = Some((i, j));
                break;
            }
        }
    }

    let part2 = if let Some((i, j)) = final_pair {
        coords[i].0 * coords[j].0
    } else {
        0
    };

    (part1.to_string(), part2.to_string())
}

fn main() {
    let data = fs::read_to_string("../data/input.txt").expect("Error reading file");
    let (part1, part2) = solve(&data);
    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
